src/Feeds.js

```diff
@@ -1,270 +1,425 @@
+      if (field.type && (field.type !== '') && !SlashtagsFeeds.VALID_TYPES.includes(field.type)) {
+        throw new SlashtagsFeeds.Error(SlashtagsFeeds.err.badFieldType)
+      }
+    })
+  }
+
+  static generateSchema (config) {
+    const { schemaConfig } = config
+    SlashtagsFeeds.validateSchemaConfig(schemaConfig)
+
+    const schema = {
+      name: schemaConfig.name,
+      description: schemaConfig.description,
+      type: 'account_feed',
+      version: '0.0.1',
+      icons: {}
+    }
+
+    for (const size in schemaConfig.icons) {
+      schema.icons[size] = schemaConfig.icons[size]
+    }
+
+    schema.fields = schemaConfig.fields.map((field) => {
+      return {
+        name: field.name,
+        description: field.description,
+        main: SlashtagsFeeds.getFileName(field),
+        type: field.type || 'utf-8'
+      }
+    })
+
+    return schema
+  }
+
+  static persistSchema (schema) {
+    fs.writeFileSync(this.DEFAULT_SCHEMA_PATH, Buffer.from(JSON.stringify(schema, undefined, 2)), 'utf-8')
+  }
+
+  static getFileName (field) {
+    const regex = /[^a-z0-9]+/gi
+    const trailing = /-+$/
+
+    return `/${field.name.toLowerCase().trim().replace(regex, '-').replace(trailing, '')}/`
+  }
+
   /**
-   *
    * @param {String} config.db.name Database name
    * @param {String} config.db.path Database path location
    * @param {String} config.slashtags.path Feeds storage path location
    * @param {String} config.slashtags.key Feeds seed key
    */
   constructor (config) {
-    this.config = config
-    this.db = new UserDb(config.db)
-    this.feed_schema = config.feed_schema
+    // Either schema needs to provided or its configuration
+    if (config.schemaConfig && !config.feed_schema) throw new Err(_err.badConfig)
     if (!config.slashtags) throw new Err(_err.badConfig)
-    this.validateFeed(this.feed_schema)
+
+    let feedSchema
+    if (config.schemaConfig) {
+      feedSchema = SlashtagsFeeds.generateSchema(config)
+      SlashtagsFeeds.persistSchema(feedSchema)
+    } else if (config.feed_schema) {
+      feedSchema = config.feed_schema
+    }
+    SlashtagsFeeds.validateSchemaConfig(feedSchema)
+
+    this.config = config
+    this.db = new FeedDb(config.db)
+    this.feed_schema = feedSchema
     this.ready = false
     this.slashtags = null
     this.lock = new Map()
   }
 
-  validateFeed (schema) {
-    if (!schema) throw new Err(_err.invalidSchema)
-    const keys = [
-      'image', 'name', 'feed_type', 'version'
-    ]
-    keys.forEach((k) => {
-      if (!schema[k]) throw new Err(_err.invalidSchema)
-    })
-  }
-
   async start () {
     try {
       await this.db.init()
     } catch (err) {
-      console.log(err)
       log.err(err)
       throw new Err(_err.dbFailedStart)
     }
-    this.slashtags = new SlashtagsFeedsLib(this.config.slashtags, this.feed_schema)
+    this.slashtags = new Feeds(this.config.slashtags, this.feed_schema)
```

Comment:
`this.slashtags` is a confusing name, given that there is no identity here at all, just hyperdrives.

Maybe `this._feeds` or `this._slashfeeds` or `this._feedsBuilder` or something.

Reply:
agreed, any reasoning for name starting with `_`?

Reply:
I conservatively name everything with `_` to indicate it is a private variable and could change at any minor release.

I try to keep public variables to a minimum, and only variables I know consumers will need and won't change between minor releases.